byzcoin/struct.go

```diff
@@ -740,3 +741,75 @@ const (
 	// instances. If not present, missing instances are ignored.
 	GUFSendMissingProofs
 )
+
+// runSingleWG allows to ensure a single execution of a method.
+type runSingleWG struct {
+	sync.WaitGroup
```

Comment:
I don't think you need to bundle a `sync.WaitGroup` as `running` is already doing the whole work.

Reply:
That's in a later PR where I need to `runSingleWG.Wait()` to be sure the catchup is done.

Reply:
Okay, makes sense then. So I'd recommend to avoid exposing `sync.WaitGroup` direclty now, as you define `done` which doesn't override `sync.WaitGroup.Done`, one can go negative by running the latter before the former. And one can `sync.WaitGroup.Add` anything.

Reply:
Btw, a negative WG-counter panics: https://play.golang.org/p/H-jQb6HAgs4